### PR TITLE
Offsettool

### PR DESCRIPTION
**Pull request by 3nids**
Offset tool was completely broken.
This is a small refactoring of the code.
As a bonus, there is now advanced config in the user input widget too.

feb-12-2018 22-57-01

### CODE REVIEW

**Comment by nirvn** [comment]
@3nids , nice. Does it maintain support for polygons? And does it maintain support for handling individual parts in a multipart geomtry?

**Comment by 3nids** [comment]
yes it should.
I will still test all combinations tomorrow (only tested single line up to now).

**Comment by nyalldawson** [comment]
Much improved! Thank you - it was quite broken before.

File: src/app/qgsmaptooloffsetcurve.cpp [hunk 1]
```diff
@@ -416,48 +387,129 @@ void QgsMapToolOffsetCurve::deleteRubberBandAndGeometry()
...
+// ******************
+// Offset User Widget
+
+QgsOffsetUserWidget::QgsOffsetUserWidget( QWidget *parent )
+  : QWidget( parent )
+{
+  setupUi( this );
+
+  // fill comboboxes
+  mJoinStyleComboBox->addItem( tr( "round" ), QgsGeometry::JoinStyleRound );
```
Comment (nyalldawson) [inline src/app/qgsmaptooloffsetcurve.cpp]: Could you change these items to be capitalized, ie. "Round"

**Comment by nyalldawson** [comment]
One other thing - better make that spin box have at least 6 digits, or it's not gonna play nice with geographic CRS

File: src/ui/qgsoffsetuserinputwidget.ui [hunk 2]
```diff
@@ -0,0 +1,169 @@
...
+  <include location="../../images/images.qrc"/>
+  <include location="../../images/images.qrc"/>
+  <include location="../../images/images.qrc"/>
+  <include location="../../images/images.qrc"/>
+  <include location="../../images/images.qrc"/>
+  <include location="../../images/images.qrc"/>
+  <include location="../../images/images.qrc"/>
+  <include location="../../images/images.qrc"/>
+  <include location="../../images/images.qrc"/>
+  <include location="../../images/images.qrc"/>
```
Comment (DelazJ) [inline src/ui/qgsoffsetuserinputwidget.ui]: are all these include normal?

File: src/app/qgsmaptooloffsetcurve.h [hunk 3]
```diff
@@ -36,15 +63,23 @@ class APP_EXPORT QgsMapToolOffsetCurve: public QgsMapToolEdit
...
+
+    void cancel();
 
   private:
     //! Rubberband that shows the position of the offset curve
     QgsRubberBand *mRubberBand = nullptr;
+    //! Snapping indicators
+    std::unique_ptr<QgsSnapIndicator> mSnapIndicator;
+
+    //! The layer maniuplated
```
Comment (DelazJ) [inline src/app/qgsmaptooloffsetcurve.h]: (being?) **manipulated**

File: src/ui/qgsoffsetuserinputwidget.ui (hunk 2 above)
Comment (nyalldawson) [reply src/ui/qgsoffsetuserinputwidget.ui]: It's caused by a bug in qtcreator. Everyone's learnt to ignore it :p

**Comment by 3nids** [comment]
@nirvn @nyalldawson thanks for your comments, they have been adressed.

Tested for (multi) lines and polygons.

Only thing, the keyboard modifier is CTRL to add instead of moving. But on mac this leads to a right-click....?

**Pull request merged**